src/server/server_family.cc

```diff
@@ -535,23 +535,40 @@ void ClientKill(CmdArgList args, absl::Span<facade::Listener*> listeners, SinkRe
 
   const bool is_admin_request = cntx->conn()->IsPrivileged();
 
+  std::vector<std::vector<facade::Connection::WeakRef>> thread_connections(pp->size());
+
   atomic<uint32_t> killed_connections = 0;
   atomic<uint32_t> kill_errors = 0;
-  auto cb = [&](unsigned thread_index, util::Connection* conn) {
+
+  auto traverse_cb = [&](unsigned idx, util::Connection* conn) {
     facade::Connection* dconn = static_cast<facade::Connection*>(conn);
     if (evaluator(dconn)) {
       if (is_admin_request || !dconn->IsPrivileged()) {
-        dconn->ShutdownSelf();
-        killed_connections.fetch_add(1);
+        thread_connections[idx].push_back(dconn->Borrow());
       } else {
         kill_errors.fetch_add(1);
       }
     }
   };
 
-  for (auto* listener : listeners) {
-    listener->TraverseConnections(cb);
-  }
+  auto cb = [&](unsigned idx, ProactorBase* p) mutable {
+    // Step 1 aggregate the per thread connections from all listeners
+    std::vector<facade::Connection::WeakRef> connections;
+    for (auto* listener : listeners) {
+      listener->TraverseConnectionsOnThread(traverse_cb, UINT32_MAX, nullptr);
+    }
+
+    // Step 2 kill the clients
+    for (auto& tcon : thread_connections[idx]) {
+      facade::Connection* conn = tcon.Get();
+      if (conn && conn->socket()->proactor()->GetPoolIndex() == p->GetPoolIndex()) {
```

Comment:
@romange 

The problem here is that we might migrate a connection between preemptions of Shutdown. Calling `tcon.Get()` on a connection that was migrated can become a data race.


Maybe the solution is to simply disable the migrations for each connection we are about to kill ? WDYT ?

Reply:
And now I am thinking about it, when we call `Borrow()` we should disable migrations. If a connection migrates after borrow, we cached `socket_->proactor()->GetPoolIndex()` within `ConnectionWeakRef` which is no longer true. The DCHECK() within `Get()` won't protect us either, it uses the cached yet no longer valid proactor index

Reply:
we run with `migrate_connections=false` in production so it's not an issue.
i think your checks are good enough for this.